core/drivers/imx_ele.c

```diff
@@ -0,0 +1,579 @@
+	rsp.status = (word & GENMASK_32(7, 0)) >> 0;
+
+	return rsp;
+}
+
+/*
+ * Initiate a communication with the EdgeLock Enclave. It sends a message
+ * and expects an answer.
+ *
+ * @msg MU message
+ */
+static TEE_Result imx_ele_call(struct imx_mu_msg *msg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct response_code rsp = { };
+	vaddr_t va = 0;
+
+	assert(msg);
+
+	if (msg->header.tag != ELE_REQUEST_TAG) {
+		EMSG("Request has invalid tag: %" PRIx8 " instead of %" PRIx8,
+		     msg->header.tag, ELE_REQUEST_TAG);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	va = imx_ele_init(ELE_BASE_ADDR, ELE_BASE_SIZE);
+	if (!va) {
+		EMSG("Fail to get base address");
+		return TEE_ERROR_GENERIC;
+	}
+
+	res = imx_mu_call(va, msg, true);
+	if (res) {
+		EMSG("Failed to transmit message: %" PRIx32, res);
+		print_msg_header(msg->header);
+		dump_message(msg);
+		return res;
+	}
+
+	rsp = get_response_code(msg->data.u32[0]);
+
+	if (msg->header.tag != ELE_RESPONSE_TAG) {
+		EMSG("Response has invalid tag: %" PRIx8 " instead of %" PRIx8,
+		     msg->header.tag, ELE_RESPONSE_TAG);
+		print_msg_header(msg->header);
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (rsp.status == ELE_COMMAND_FAILED) {
+		EMSG("Command has failed");
+		print_rsp_code(rsp);
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* The rating can be different in success and failing case */
+	if (rsp.rating != 0) {
+		EMSG("Command has invalid rating: %" PRIx8, rsp.rating);
+		print_rsp_code(rsp);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Get device information from EdgeLock Enclave
+ *
+ * @session_handle EdgeLock Enclave session handler
+ * @rsp Device info
+ */
+static TEE_Result
+imx_ele_session_get_device_info(uint32_t session_handle,
+				struct session_get_device_info_rsp *rsp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	struct session_get_device_info_cmd {
+		uint32_t session_handle;
+	} cmd = {
+		.session_handle = session_handle,
+	};
+
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_HSM,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_SESSION_DEVICE_INFO,
+	};
+
+	assert(rsp);
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	res = imx_ele_call(&msg);
+	if (res)
+		return res;
+
+	memcpy(rsp, msg.data.u32, sizeof(*rsp));
+
+	if (compute_crc(&msg) != rsp->crc)
+		return TEE_ERROR_CORRUPT_OBJECT;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Open a session with EdgeLock Enclave. It return a session handler.
+ *
+ * @session_handle EdgeLock Enclave session handler
+ */
+static TEE_Result imx_ele_session_open(uint32_t *session_handle)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct open_session_cmd {
+		uint8_t mu_id;
+		uint8_t interrupt_num;
+		uint8_t tz;
+		uint8_t did;
+		uint8_t priority;
+		uint8_t op_mode;
+		uint16_t reserved;
+	} __packed cmd = {
+		.mu_id = ELE_MU_ID,
+		.interrupt_num = ELE_MU_IRQ,
+		.tz = 0,
+		.did = ELE_MU_DID,
+		.priority = 0,
+		.op_mode = 0,
+		.reserved = 0,
+	};
+
```

Comment:
Minor: this blank line is not needed, all variable are normally declared in one block. Same for the next one and below.

Reply:
Ok, I fixed all functions